[SDR-259] BE : 리뷰 작성 시 선택한 가게의 상세 주소를 카카오 로컬 API 에서조회하여 저장하도록 변경

jinan159 opened the pull request:
### ❗️ 이슈 번호

[SDR-259]

<br><br>

### 📝 구현 내용

- 가게 등록 검증 시, 카카오 API 를 통해 조회한 가게의 상세 주소를 조회하는 로직 추가
  - StoreService 에서 상세 주소 조회 실패 시, `Address` 의 필수값만 저장하도록 함(addressName, roadAddressName)
    카카오 API 가게 정보 조회에서 addressName, roadAddressName 두 정보가 있어서 저장 가능
<img width="595" alt="image" src="https://user-images.githubusercontent.com/45728407/188815679-8f832510-799e-4f88-a02d-e241ac51617b.png">
<img width="431" alt="image" src="https://user-images.githubusercontent.com/45728407/188815820-6f85e09b-8c47-400c-b883-dfbe56304691.png">

<br><br>

### 🙇🏻‍♂️ 리뷰어에게 부탁합니다!

- (리뷰어에게 부탁하는 말을 작성해주세요. 없다면 지워도 됩니다!)

<br><br>

### 💡 참고 자료

- (없다면 지워도 됩니다!)


[SDR-259]: https://jjikmuk.atlassian.net/browse/SDR-259?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

jinan159 committed:
SDR-259: docs: 카카오 로컬 API 장소, 주소검색 응답 json 추가

jinan159 committed:
SDR-259: test: 테스트 resources 를 다루는 유틸 클래스 추가

jinan159 committed:
SDR-259: test: PlaceMocksExtention 에서 KakaoPlaceMocks 의 모든 Mock API 를 시작, 리셋 하도록 변경

jinan159 committed:
SDR-259: test: KakaoPlaceMocks reset 로직 오류 수정

- name 으로 종료하고 있었는데, id 로 종료(reset)해야함

jinan159 committed:
SDR-259: feat: 카카오 로컬 주소 조회 API 기능 추가 및 통합테스트 통과

jinan159 committed:
SDR-259: test: 카카오 로컬 API Mock 리펙토링

- 시나리오 없이 stub 설정되도록 변경
- stub 초기화 로직 개선

jinan159 committed:
SDR-259: test: 테스트 이름 변경

jinan159 committed:
SDR-259: feat: 가게 등록 검증 API 주소 상세 저장로직 구현 및 통합테스트 통과

jinan159 committed:
SDR-259: docs: 가게 등록 검증 API 문서 추가

ku-kim commented on be/src/main/java/com/jjikmuk/sikdorak/store/domain/Address.java:
```diff
@@ -3,12 +3,16 @@
 
 @Getter
 @Embeddable
+@AllArgsConstructor
 @NoArgsConstructor(access = AccessLevel.PROTECTED)
+@Builder
```
변수가 많아져서 빌더와 allargs~ 를 사용했군요!

- 안전한 빌더 패턴(null 체크)이 아니라서 DB를 거쳐야 에러가 발생하는지 알 거 같아요

- 필드가 모두 String 이라서, 만약 필드 위치가 바뀐다면 AllArgsConstructor 사용했을 때 문제를 찾지 못할 거 같아요.

어떻게 생각하세요?

---

ref : 안전한 빌더패턴 https://jaehoney.tistory.com/229

jinan159 replied:
도메인 관련 객체라, 필수값을 생성자에서 체크하는 형태로 가야겠군요..!
이 부분은 개선해보겠습니다!

ku-kim commented on be/src/main/java/com/jjikmuk/sikdorak/store/service/dto/AddressSearchResponse.java:
```diff
@@ -0,0 +1,17 @@
+import lombok.NoArgsConstructor;
+
+@NoArgsConstructor
+@AllArgsConstructor
```
p3; 여기는 AllArgsConstructor을 사용하고있어요

jinan159 replied:
DTO 들 생성자 싹 다 정리해서 다시 푸시할게요!

seokho-ham commented on be/src/test/java/com/jjikmuk/sikdorak/common/TestResourceUtils.java:
```diff
@@ -0,0 +1,22 @@
+
+public abstract class TestResourceUtils {
+
+	public static String getTestResourceWithPath(String path) {
+		try {
+			return copyToString(getResourceStreamWithPath(path), defaultCharset());
+		} catch (IOException e) {
+			throw new RuntimeException(e);
+		}
+	}
+
```
여러곳에서 반복되는 코드를 Util로 아예 빼는 선택 좋은거 같습니다!!👍

ku-kim commented on be/src/main/java/com/jjikmuk/sikdorak/store/service/dto/AddressSearchRequest.java:
```diff
@@ -0,0 +1,15 @@
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+
+@Getter
+@NoArgsConstructor
+public class AddressSearchRequest {
+
+	private String query;
+
+	public AddressSearchRequest(String query) {
+		this.query = query;
+	}
+}
```
p2; 여기는 생성자를 사용하고 있네요! 통일이 필요해보여요

+a) `sikdorak/store/service/dto/*` 하위 객체들 모두 체크가 필요해요

seokho-ham commented on be/src/test/java/com/jjikmuk/sikdorak/integration/store/StoreVerifyOrSaveIntegrationTest.java:
```diff
@@ -2,76 +2,246 @@
+			.thenReturn(getAddressSearchResponse());
+	}
 
 	@Nested
```
p99. 단위 테스트에서만 계층구조로 테스트를 작성하기로 했던것 같은 기억이.. (갑자기 눈에 들어와서 남겨요! 바꿔야한다는건 아닙니다ㅎㅎ)

ku-kim replied:
통합 테스트에도 적용하기 괜찮은 구조라면 써도 좋다고 전에 이야기 했던거같아요!

seokho-ham replied:
아 그렇군요! 제 기억이 잘못되었던걸로!

seokho-ham commented on be/src/test/java/com/jjikmuk/sikdorak/integration/store/StoreVerifyOrSaveIntegrationTest.java:
```diff
@@ -2,76 +2,246 @@
+				assertThat(address.getRegion2DepthName()).isEqualTo(expectedRegion2DepthName);
+				assertThat(address.getRegion3DepthName()).isEqualTo(expectedRegion3DepthName);
+				assertThat(address.getRegion3DepthHName()).isEqualTo(expectedRegion3DepthHName);
+			}
```
p3. assertAll로 묶는것도 나쁘지 않아 보입니다!

seokho-ham commented on be/src/test/java/com/jjikmuk/sikdorak/integration/store/StoreVerifyOrSaveIntegrationTest.java:
```diff
@@ -2,76 +2,246 @@
+				assertThat(address.getRegion2DepthName()).isNull();
+				assertThat(address.getRegion3DepthName()).isNull();
+				assertThat(address.getRegion3DepthHName()).isNull();
+			}
```
p3. 여기도요!

seokho-ham commented on be/src/main/java/com/jjikmuk/sikdorak/store/service/KakaoPlaceApiService.java:
```diff
@@ -49,4 +59,45 @@ private static List<PlaceResponse> convertToPlaceResponseList(KakaoPlaceSearchRe
 	}
+
+	@Override
+	public AddressSearchResponse searchAddress(AddressSearchRequest searchRequest) {
+		String addressName = searchRequest.getQuery();
+		validateAddressName(addressName);
+
+		KakaoAddressSearchResponse addressResponse = addressSearchClient.searchAddress(addressName);
+
+		return new AddressSearchResponse(convertToAddressResponseList(addressResponse));
+	}
+
```
p2. 이 메서드를 위에 있는 private 메서드보다 위로 올리는게 좋을것 같습니다!

seokho-ham commented on be/src/main/java/com/jjikmuk/sikdorak/store/service/client/KakaoLocalAddressSearchClient.java:
```diff
@@ -0,0 +1,16 @@
+	FeignClientOAuthErrorConfiguration.class})
+public interface KakaoLocalAddressSearchClient {
+
+	@GetMapping("/v2/local/search/address.json")
```
카카오 api path가 신기하네요;

jinan159 replied:
출력 포멧을 json, xml 을 지원하는데, 끝에 확장자 처럼 붙여서 사용하게 되어있더라구요!
신박한거같습니다 ㅋㅋ

seokho-ham reviewed (dismissed):
고생하셨어요 제이!

ku-kim commented on be/src/test/java/com/jjikmuk/sikdorak/integration/store/PlaceApiIntegrationTest.java:
```diff
@@ -86,4 +88,28 @@ static Stream<Arguments> providePointInvalidPair() {
+			AddressSearchResponse responses = kakaoPlaceApiService.searchAddress(request);
+
+			// then
+			assertThat(responses.getAddressResponses()).isNotEmpty()
+				.satisfies(addressList -> {
+					addressList.forEach(address -> {
+						assertThat(address.addressName()).isEqualTo(query);
+					});
+				});
```
p3; 코드 개선할 수 있어요.

all, any, noneSatisfy() 사용하면 List를 이미 풀린 상태(객체)에서 쓸 수 있습니다.

```suggestion
			assertThat(responses.getAddressResponses()).isNotEmpty()
				.anySatisfy(address -> assertThat(address.addressName()).isEqualTo(query));
```


ref : https://umanking.github.io/2021/06/26/assertj-iteration/

ku-kim reviewed (changes requested):
고생 많으셨어요~ jay!

jinan159 committed:
SDR-259: refactor: 카카오 API 응답 클래스 생성자 정리

- 불필요한 lombok 생성자 어노테이션 제거
- List 반환시, unmodifiableList 로 반환하도록 변경
  - lombok Getter 제거

jinan159 committed:
SDR-259: refactor: 장소 API 응답 클래스 생성자 리펙토링

- @AllArgsConstructor 어노테이션 제거
- 필드 final 로 변경

jinan159 committed:
SDR-259: refactor: Address 생성자, Builder 리펙토링

- 빌더 패턴 -> 안전한 빌더 패턴
- 정적 팩토리 메소드를 안전한 빌더 패턴으로 대체
- @AllArgsConstructor 제거
  - 생성자에서 필수값 검증로직 추가

jinan159 committed:
SDR-259: test: 테스트 코드 검증로직 개선

- StoreVerifyOrSaveIntegrationTest : assertAll 활용 변경
- PlaceApiIntegrationTest : allSatisfy 활용하도록 변경

seokho-ham commented on be/src/main/java/com/jjikmuk/sikdorak/store/domain/Address.java:
```diff
@@ -59,10 +59,32 @@ public class Address {
+    }
+
+    public static AddressBuilder requiredFieldBuilder(String addressName, String roadAddressName) {
+        return builder()
+            .addressName(addressName)
+            .roadAddressName(roadAddressName);
```
메서드명 👍👍

seokho-ham reviewed (dismissed):
제이 고생하셨어요! 👍

ku-kim commented on be/src/main/java/com/jjikmuk/sikdorak/store/domain/Address.java:
```diff
@@ -1,14 +1,18 @@
 @Getter
 @Embeddable
 @NoArgsConstructor(access = AccessLevel.PROTECTED)
+@Builder
```
```suggestion
@Builder(builderMethodName = "hiddenBuilder")
```

builder() 메서드가 외부에서 노출되니 이름으로 사용하지 말라고 명시해주는건 어때요?

ku-kim replied:
hiddenBuilder() 빌더를 아예 없애고 싶었지만, 

아래 예시에서 나온 `@Builder(builderMethodName = "hiddenBuilder", access = AccessLevel.PRIVATE)` 은 작동하지 않더라구요. 제 생각엔 이게 최선인거같아요

https://stackoverflow.com/questions/29885428/required-arguments-with-a-lombok-builder

ku-kim commented on be/src/main/java/com/jjikmuk/sikdorak/store/domain/Address.java:
```diff
@@ -57,12 +77,14 @@ public Address(String addressName, String roadAddressName, String region1DepthNa
-            null, null, null, null, null, null
-        );
+    public static AddressBuilder requiredFieldBuilder(String addressName, String roadAddressName) {
+        return builder()
```
```suggestion
        return hiddenBuilder()
```

ku-kim replied:
- 적용전

![image](https://user-images.githubusercontent.com/57086195/189073204-5d4eae45-e428-4556-86b3-b6424d3134c6.png)

- 적용 후
![image](https://user-images.githubusercontent.com/57086195/189073430-09d3ba45-0cb0-4e44-b9f7-fe0a7e2dac50.png)

jinan159 committed:
SDR-259: build: 배포 관련 설정 최신화

jinan159 committed:
SDR-259: refactor: 기본 빌더 사용을 자제하도록 하기 위해 명칭 변경
